crates/ruff/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,964 @@
+            return; // Block already visited, no needed to do it again.
+        }
+
+        match &block.next {
+            NextBlock::Always(next) => idx = *next,
+            NextBlock::If {
+                condition,
+                next,
+                orelse,
+            } => {
+                match taken(condition) {
+                    Some(true) => idx = *next,    // Always taken.
+                    Some(false) => idx = *orelse, // Never taken.
+                    None => {
+                        // Don't know, both branches might be taken.
+                        idx = *next;
+                        mark_reached(reached_map, blocks, *orelse);
+                    }
+                }
+            }
+            NextBlock::Terminate => return,
+        }
+    }
+}
+
+/// Determines if `condition` is taken.
+/// Returns `Some(true)` if the condition is always true, e.g. `if True`, same
+/// with `Some(false)` if it's never taken. If it can't be determined it returns
+/// `None`, e.g. `If i == 100`.
+fn taken(condition: &Condition) -> Option<bool> {
+    // TODO: add more cases to this where we can determine a condition
+    // statically. For now we only consider constant booleans.
+    match condition {
+        Condition::Test(expr) => match expr {
+            Expr::Constant(constant) => constant.value.as_bool().copied(),
+            _ => None,
+        },
+        Condition::Iterator(_) => None,
+        Condition::Match { .. } => None,
+    }
+}
+
+/// Collection of basic block.
+#[derive(Debug, PartialEq)]
+struct BasicBlocks<'stmt> {
+    /// # Notes
+    ///
+    /// The order of these block is unspecified. However it's guaranteed that
+    /// the last block is the statement in the function and the first block is
+    /// the last statement. The block are more or less in reverse order, but it
+    /// gets fussy around control flow statements (e.g. `if` statements).
+    ///
+    /// For loop blocks, and similar recurring control flows, the end of the
+    /// body will point to the loop block again (to create the loop). However an
+    /// oddity here is that this block might contain statements before the loop
+    /// itself which, of course, won't be executed again.
+    ///
+    /// For example:
+    /// ```python
+    /// i = 0          # block 0
+    /// while True:    #
+    ///     continue   # block 1
+    /// ```
+    /// Will create a connection between block 1 (loop body) and block 0, which
+    /// includes the `i = 0` statement.
+    ///
+    /// To keep `NextBlock` simple(r) `NextBlock::If`'s `next` and `orelse`
+    /// fields only use `BlockIndex`, which means that they can't terminate
+    /// themselves. To support this we insert *empty*/fake blocks before the end
+    /// of the function that we can link to.
+    ///
+    /// Finally `BasicBlock` can also be a sentinel node, see the associated
+    /// constants of [`BasicBlock`].
+    blocks: Vec<BasicBlock<'stmt>>,
```

Comment:
Nit: We can use [`IndexVec`](https://github.com/astral-sh/ruff/blob/dd8fc4876ae590e0d906e05717b7a68c3e1a3244/crates/ruff_index/src/vec.rs#L11) here

Reply:
I tried doing this, but `ruff` doesn't seem to dependent on `ruff_index` currently. Is is worth added the dependency just for this one type?